File: Arrays_And_Hashing/PythonSolutions/WordPattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        
        char_to_word={}
        word_to_char={}
        words = s.split()

        if len(pattern) != len(words):
            return False


        for char,word in zip(pattern,words):
            if char not in char_to_word and word not in word_to_char:
                char_to_word[char] = word
                word_to_char[word] = char

            if char_to_word.get(char)!=word or word_to_char.get(word)!=char:
                return False
        return True

File: Arrays_And_Hashing/JavaSolutions/WordPattern.java
package Arrays_And_Hashing.JavaSolutions;

import java.util.*;

class WordPattern{
    public boolean wordPattern(String pattern, String s) {
        String[] words = s.split(" ");

        if (pattern.length() != words.length) {
            return false;
        }

        Map<Character, String> charToWord = new HashMap<>();
        Map<String, Character> wordToChar = new HashMap<>();

        for (int i = 0; i < pattern.length(); i++) {
            char c = pattern.charAt(i);
            String word = words[i];

            if (!charToWord.containsKey(c) && !wordToChar.containsKey(word)) {
                charToWord.put(c, word);
                wordToChar.put(word, c);
            } else {
                if (!word.equals(charToWord.get(c)) || wordToChar.get(word) != c) {
                    return false;
                }
            }
        }

        return true;
    }
}
